circom/circom_test.go

```diff
@@ -0,0 +1,338 @@
+package circom
+
+import (
+	"fmt"
+	"github.com/consensys/gnark-crypto/ecc"
+	fr_bn254 "github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	"github.com/consensys/gnark-crypto/ecc/secp256k1"
+	"github.com/consensys/gnark-crypto/ecc/secp256k1/fp"
+	fr_secp "github.com/consensys/gnark-crypto/ecc/secp256k1/fr"
+	groth16 "github.com/consensys/gnark/backend/groth16/bn254"
+	"github.com/consensys/gnark/backend/groth16/bn254/mpcsetup"
+	"github.com/consensys/gnark/constraint"
+	cs "github.com/consensys/gnark/constraint/bn254"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/consensys/gnark/std/algebra/emulated/sw_emulated"
+	zksha3 "github.com/consensys/gnark/std/hash/sha3"
+	"github.com/consensys/gnark/std/math/emulated"
+	"github.com/consensys/gnark/std/math/uints"
+	"github.com/consensys/gnark/test"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"golang.org/x/crypto/sha3"
+	"math/big"
+	"math/rand"
+	"testing"
+	"time"
+)
+
+func doMPCSetUp(ccs constraint.ConstraintSystem) (pk groth16.ProvingKey, vk groth16.VerifyingKey, err error) {
+	const (
+		nContributionsPhase1 = 3
+		nContributionsPhase2 = 3
+		power                = 20 //19 //2^9 元素个数
+	)
+
+	srs1 := mpcsetup.InitPhase1(power)
+
+	// Make and verify contributions for phase1
+	for i := 1; i < nContributionsPhase1; i++ {
+		// we clone test purposes; but in practice, participant will receive a []byte, deserialize it,
+		// add his contribution and send back to coordinator.
+		prev := Phase1clone(srs1)
+
+		srs1.Contribute()
+		err := mpcsetup.VerifyPhase1(&prev, &srs1)
+		if err != nil {
+			return groth16.ProvingKey{}, groth16.VerifyingKey{}, err
+		}
+	}
+
+	// Prepare for phase-1.5
+	var evals mpcsetup.Phase2Evaluations
+	r1cs := ccs.(*cs.R1CS)
+	// Prepare for phase-2
+	srs2, evals := mpcsetup.InitPhase2(r1cs, &srs1)
+	// Make and verify contributions for phase1
+	for i := 1; i < nContributionsPhase2; i++ {
+		// we clone for test purposes; but in practice, participant will receive a []byte, deserialize it,
+		// add his contribution and send back to coordinator.
+		prev := Phase2clone(srs2)
+
+		srs2.Contribute()
+		err := mpcsetup.VerifyPhase2(&prev, &srs2)
+		if err != nil {
+			return groth16.ProvingKey{}, groth16.VerifyingKey{}, err
+		}
+	}
+
+	// Extract the proving and verifying keys
+	pk, vk = mpcsetup.ExtractKeys(&srs1, &srs2, &evals, ccs.GetNbConstraints())
+	return pk, vk, nil
+}
+
+func Phase1clone(phase1 mpcsetup.Phase1) mpcsetup.Phase1 {
+	r := mpcsetup.Phase1{}
```

Comment:
If new fields are added to Phase1 or Phase2, these functions might not clone them, is this expected?

Reply:
some here looks like to be shalom

Reply:
Yes. The purpose here is to create a copy of the current state. Don't expect subsequent modifications to change it